### PR TITLE
Disable JIT engine on macOS platforms by default

### LINKED ISSUE #8252: Disable JIT engine on macOS platforms by default
There is a bunch of tickets related to LuaJIT compiler on macOS platforms. To improve customer experience me and @Mons decided to disable JIT engine on Tarantool startup for macOS builds. Either way, JIT will be aboard as a result of the changes and more adventurous users will be able to enable it via `jit.on()` in their code.

The following need to be done in scope of this issue:
* [x] Adopt LuaJIT testing machinery to skip JIT-related tests
* [x] Introduce LuaJIT CI workflow for regular testing routine with disabled compiler
* [ ] Implement the patch disabling JIT machinery in Tarantool on macOS
* [ ] Adopt Tarantool testing suite to skip JIT-related tests
* [ ] Enable M1 regular testing back.

### PR DESCRIPTION
**Pull request by igormunkin**
This PR contains the following changes:

1. Several Tarantool tests are adjusted (and even rewritten) to the Tarantool builds with JIT being turned off.
2. LuaJIT compiler is turned off by default on macOS and `LUAJIT_JIT_STATUS` CMake option is introduced
3. Testing routine is enabled back for macOS/M1 workflows in Tarantool CI.

Closes #8252

### CODE REVIEW

File: .test.mk [hunk 1]
```diff
@@ -172,7 +172,7 @@ test-osx-release: prebuild-osx build run-luajit-test pretest-osx run-test
 test-osx-release-arm64: CMAKE_PARAMS = -DCMAKE_BUILD_TYPE=RelWithDebInfo \
                                        -DENABLE_WERROR=ON \
                                        -DTEST_BUILD=ON
-test-osx-release-arm64: prebuild-osx build
+test-osx-release-arm64: prebuild-osx build run-luajit-test pretest-osx run-test
```
Comment (ylobankov) [inline .test.mk:175]: Please remove this target at all. You have enabled testing, so you can use `test-osx-release` target. The same for debug and static builds. Also, please fix all macOS workflows: delete this logic https://github.com/tarantool/tarantool/blob/9549310d15dd16c6fd49cc658bdf9fffb82ba63c/.github/workflows/osx_debug.yml#L61-L62.

Comment (igormunkin) [reply .test.mk]: @ylobankov, thanks for your review! I've fixed all spots you've mentioned above.

```diff
diff --git a/.github/workflows/osx_debug.yml b/.github/workflows/osx_debug.yml
index 515dc51ec..a53e0c678 100644
--- a/.github/workflows/osx_debug.yml
+++ b/.github/workflows/osx_debug.yml
@@ -58,9 +58,7 @@ jobs:
       - name: Install deps
         uses: ./.github/actions/install-deps-osx
       - name: test
-        env:
-          SUFFIX: ${{ matrix.arch == 'aarch64' && '-arm64' || '' }}
-        run: make -f .test.mk test-osx-debug${SUFFIX}
+        run: make -f .test.mk test-osx-debug
       - name: Send VK Teams message on failure
         if: failure()
         uses: ./.github/actions/report-job-status
diff --git a/.github/workflows/osx_release.yml b/.github/workflows/osx_release.yml
index 425bc96b1..fb18179ab 100644
--- a/.github/workflows/osx_release.yml
+++ b/.github/workflows/osx_release.yml
@@ -58,9 +58,7 @@ jobs:
       - name: Install deps
         uses: ./.github/actions/install-deps-osx
       - name: test
-        env:
-          SUFFIX: ${{ matrix.arch == 'aarch64' && '-arm64' || '' }}
-        run: make -f .test.mk test-osx-release${SUFFIX}
+        run: make -f .test.mk test-osx-release
       - name: Send VK Teams message on failure
         if: failure()
         uses: ./.github/actions/report-job-status
diff --git a/.github/workflows/osx_release_lto.yml b/.github/workflows/osx_release_lto.yml
index 0192e1c2b..f81139316 100644
--- a/.github/workflows/osx_release_lto.yml
+++ b/.github/workflows/osx_release_lto.yml
@@ -60,8 +60,7 @@ jobs:
       - name: test
         env:
           CMAKE_EXTRA_PARAMS: -DENABLE_LTO=ON
-          SUFFIX: ${{ matrix.arch == 'aarch64' && '-arm64' || '' }}
-        run: make -f .test.mk test-osx-release${SUFFIX}
+        run: make -f .test.mk test-osx-release
       - name: Send VK Teams message on failure
         if: failure()
         uses: ./.github/actions/report-job-status
diff --git a/.github/workflows/osx_static_cmake.yml b/.github/workflows/osx_static_cmake.yml
index 6208f839c..a34c158e8 100644
--- a/.github/workflows/osx_static_cmake.yml
+++ b/.github/workflows/osx_static_cmake.yml
@@ -60,8 +60,7 @@ jobs:
         uses: ./.github/actions/install-deps-osx
       - name: test
         env:
-          SUFFIX: ${{ matrix.arch == 'aarch64' && '-arm64' || '' }}
-        run: make -f .test.mk test-osx-static-cmake${SUFFIX}
+        run: make -f .test.mk test-osx-static-cmake
       - name: Send VK Teams message on failure
         if: failure()
         uses: ./.github/actions/report-job-status
diff --git a/.test.mk b/.test.mk
index d2f80c430..5b9e884cf 100644
--- a/.test.mk
+++ b/.test.mk
@@ -164,16 +164,6 @@ test-osx-release: CMAKE_PARAMS = -DCMAKE_BUILD_TYPE=RelWithDebInfo \
                                  -DTEST_BUILD=ON
 test-osx-release: prebuild-osx build run-luajit-test pretest-osx run-test

-# FIXME: Temporary target without tests. Use 'test-release-osx' target instead
-# when all M1 issues are resolved:
-#   LuaJIT tests - https://github.com/tarantool/tarantool/issues/4819
-#   Tarantool tests - https://github.com/tarantool/tarantool/issues/6068
-.PHONY: test-osx-release-arm64
-test-osx-release-arm64: CMAKE_PARAMS = -DCMAKE_BUILD_TYPE=RelWithDebInfo \
-                                       -DENABLE_WERROR=ON \
-                                       -DTEST_BUILD=ON
-test-osx-release-arm64: prebuild-osx build run-luajit-test pretest-osx run-test
-
 # Debug build

 .PHONY: test-osx-debug
@@ -181,15 +171,6 @@ test-osx-debug: CMAKE_PARAMS = -DCMAKE_BUILD_TYPE=Debug \
                                -DTEST_BUILD=ON
 test-osx-debug: prebuild-osx build run-luajit-test pretest-osx run-test

-# FIXME: Temporary target without tests. Use 'test-debug-osx' target instead
-# when all M1 issues are resolved:
-#   LuaJIT tests - https://github.com/tarantool/tarantool/issues/4819
-#   Tarantool tests - https://github.com/tarantool/tarantool/issues/6068
-.PHONY: test-osx-debug-arm64
-test-osx-debug-arm64: CMAKE_PARAMS = -DCMAKE_BUILD_TYPE=Debug \
-                                     -DTEST_BUILD=ON
-test-osx-debug-arm64: prebuild-osx build run-luajit-test pretest-osx run-test
-
 # Static build

 .PHONY: test-osx-static-cmake
@@ -201,19 +182,6 @@ test-osx-static-cmake: LUAJIT_TEST_BUILD_DIR = ${STATIC_BIN_DIR}
 test-osx-static-cmake: TEST_RUN_PARAMS = --builddir ${PWD}/${STATIC_BIN_DIR}
 test-osx-static-cmake: prebuild-osx build run-luajit-test pretest-osx run-test

-# FIXME: Temporary target without tests. Use 'test-osx-static-cmake' target instead
-# when all M1 issues are resolved:
-#   LuaJIT tests - https://github.com/tarantool/tarantool/issues/4819
-#   Tarantool tests - https://github.com/tarantool/tarantool/issues/6068
-.PHONY: test-osx-static-cmake-arm64
-test-osx-static-cmake-arm64: SRC_DIR = ${STATIC_DIR}
-test-osx-static-cmake-arm64: BUILD_DIR = ${STATIC_DIR}
-test-osx-static-cmake-arm64: CMAKE_PARAMS = -DCMAKE_TARANTOOL_ARGS="-DCMAKE_BUILD_TYPE=RelWithDebInfo;-DENABLE_WERROR=ON;-DTEST_BUILD=ON"
-test-osx-static-cmake-arm64: CTEST = true
-test-osx-static-cmake-arm64: LUAJIT_TEST_BUILD_DIR = ${STATIC_BIN_DIR}
-test-osx-static-cmake-arm64: TEST_RUN_PARAMS = --builddir ${PWD}/${STATIC_BIN_DIR}
-test-osx-static-cmake-arm64: prebuild-osx build
-
 ##############################
 # FreeBSD                    #
 ##############################
```

File: test/app-luatest/lua/gh_1700_abort_recording_on_fiber_switch.lua [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+-- Set the value to trigger <yield> call switch the running fiber.
+local yield = require('libyield')(trigger)
+
+-- Depending on trigger and hotloop values the following contexts
+-- are possible:
+-- * if trigger <= hotloop -> trace recording is aborted
+-- * if trigger >  hotloop -> trace is recorded but execution
+--   leads to panic
+jit.opt.start("3", string.format("hotloop=%d", hotloop))
```
Comment (Buristan) [inline test/app-luatest/lua/gh_1700_abort_recording_on_fiber_switch.lua:16]: Minor: I suggest to use the following format to be consistent with the other code:
```lua
jit.opt.start(3, string.format('hotloop=%d', hotloop))
```
Side note: why do we set OPT level to default 3 here?

Feel free to ignore.

**Review by Buristan — COMMENTED** [review]
**[PATCH 2/8] test: rewrite test for gh-1700 with luatest**

LGTM, except 2 nits below.

> As a result of the patch all required files are moved to app-luatest

Typo: s/of the patch/of the patch,/

**Review by Buristan — COMMENTED** [review]
**[PATCH 3/8] test: rewrite test for gh-5983 with luatest**
LGTM, except a minor nit regarding the commit message. 

> subdirectory introducing the following layout:

I prefer to say "using the following layout", since it was introduced in the previous patch.
Feel free to ignore.

**Review by Buristan — COMMENTED** [review]
**[PATCH 4/8] test: drop test-run skipcond for gh-7762 test**
LGTM, as trivial.

File: cmake/luajit.cmake [hunk 3]
```diff
@@ -75,6 +75,12 @@ if(ENABLE_ASAN)
     add_definitions(-DLUAJIT_USE_ASAN=1)
 endif()
 
+if(LUAJIT_JIT_STATUS)
```
Comment (Buristan) [inline cmake/luajit.cmake:78]: I'm strongly disagree that this is the right place to put this option. <cmake/luajit.cmake> file is about __LuaJIT configuration__, not Tarantool's behaviour. I suggest to move this define into <src/trivia/config.h.cmake> like it is done for `ENABLE_BACKTRACE` and other flags.

File: CMakeLists.txt [hunk 4]
```diff
@@ -573,9 +574,17 @@ if (TARGET_OS_DARWIN)
     # LuaJIT is unusable on OS X without enabled GC64
     # See https://github.com/tarantool/tarantool/issues/2643
     set(LUAJIT_ENABLE_GC64_DEFAULT ON)
+    # To improve customer experience JIT engine is turned off on
+    # Tarantool startup for macOS builds. Either way, JIT will be
+    # aboard as a result of the changes and more adventurous users
+    # will be able to enable it via jit.on() in their code.
+    # See https://github.com/tarantool/tarantool/issues/8252
```
Comment (Buristan) [inline CMakeLists.txt:581]: Typo: s/8252/8252./

**Review by Buristan — COMMENTED** [review]
**[PATCH 5/8] lua: disable JIT engine on macOS by default**
Generally, patch LGTM, but I strongly disagree about changes in <cmake/luajit.cmake>, we should use <src/trivia/config.h.cmake> instead.

**Review by Buristan — COMMENTED** [review]
**[PATCH 6/8] test: disable app/fiber.test.lua on macOS/M1**
LGTM, except a single nit regarding the commit message.
> add skipcond condition for app/fiber.test.lua for now.

Typo: s/skipcond condition/the skipcond condition/
Also, skipcond condition sounds tautologically to me, but feel free to ignore.

**Review by Buristan — COMMENTED** [review]
**[PATCH 7/8] test: disable app-luatest/fiber_parent_backtrace_test.lua on macOS/M1**

LGTM, except a single nit regarding the commit message.

> add skip_if condition for app-luatest/fiber_parent_backtrace_test.lua

Typo: s/skip_if condition/the skip_if condition/

Also, the commit message title is longer than 50 symbols, but I suppose it's acceptable in this case.

File: test/app-luatest/lua/gh_1700_abort_recording_on_fiber_switch.lua (hunk 2 above)
Comment (igormunkin) [reply test/app-luatest/lua/gh_1700_abort_recording_on_fiber_switch.lua]: I have no idea: the default OPT level is 3 IIRC, so I just removed it. Quotes are fixed.

File: CMakeLists.txt (hunk 4 above)
Comment (igormunkin) [reply CMakeLists.txt:581]: Fixed, thanks!

**Comment by igormunkin** [comment]
Hi folks! I've fixed all the comments you've left and rebased the patchset on the new master with relevant LuaJIT bumps.

**Review by Buristan — APPROVED** [review]
Hi, Igor!
Thanks for the fixes!
LGTM!

File: .github/workflows/osx_static_cmake.yml [hunk 5]
```diff
@@ -60,8 +60,7 @@ jobs:
         uses: ./.github/actions/install-deps-osx
       - name: test
         env:
```
Comment (ylobankov) [inline .github/workflows/osx_static_cmake.yml:62]: Please remove this, otherwise, you will get a syntax error.

Comment (igormunkin) [reply .github/workflows/osx_static_cmake.yml]: Shit. Thanks for noticing! Fixed.

**Review by ylobankov — APPROVED** [review]
LGTM

**Pull request merged**